core/src/main/scala/kafka/server/ReplicaManager.scala

```diff
@@ -558,10 +558,27 @@ class ReplicaManager(val config: KafkaConfig,
     localLog(topicPartition).map(_.parentDir)
   }
 
+  // visible for testing
+  val delayedActions = new LinkedBlockingQueue[() => Unit]()
+
+  /**
+   * try to complete delayed action. In order to avoid conflicting locking, the actions to complete delayed requests
+   * are kept in a queue. We add the logic to check the ReplicaManager queue at the end of KafkaApis.handle(),
```

Comment:
at the end of KafkaApis.handle() => at the end of KafkaApis.handle() and the expiration thread for certain delayed operations (e.g. DelayedJoin)